### PR TITLE
Migrate source

### PR DESCRIPTION
**Pull request by richfitz**
This PR does a first pass attempt at migrating a source repo over from orderly 1 to orderly 3. It also includes some changes required to work with the most recent copies of outpack and orderly, which is getting increasingly hard to keep integrated as I let this branch get staler and staler (some of this work is really old; back to April at least).

Almost certainly this won't work on any really complex source repo, but it *does* work on the demo dir. There's a bit of light testing, but it's quite hard to do well in practice and I'd like some comments on the general approach before going too far.

There's also an issue with the practicalities of doing this for real, and how we cope with orderly3 being a really weird namespace floating around!

### CODE REVIEW

File: R/src.R [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+orderly2outpack_src <- function(path, delete_yml = FALSE, strict = FALSE) {
+  cfg <- orderly::orderly_config(path)
+  path <- cfg$root
+
+  nms <- orderly::orderly_list(path)
+
+  i <- file.exists(file.path(path, "src", nms, "orderly.R"))
+  if (any(i)) {
+    stop("Some source directories already contain 'orderly.R' files")
+  }
```
Comment (r-ash) [inline R/src.R:29]: Worth telling user which reports already have this file so it can be fixed?

File: tests/testthat/test-src.R [hunk 2]
```diff
@@ -0,0 +1,144 @@
+test_that("Can migrate orderly demo directory", {
+  path <- orderly_demo_src()
+  orderly2outpack_src(path, delete_yml = TRUE, strict = TRUE)
+
+  dat <- orderly:::read_demo_yml(path)
```
Comment (r-ash) [inline tests/testthat/test-src.R:5]: I think this is not testing every report can be run? In particular `default-param` is not being run

```
> reports <- vapply(dat, function(x) x$name, character(1))
> reports
 [1] "minimal"            "minimal"            "other"              "minimal"            "other"              "global"            
 [7] "use_dependency"     "other"              "other"              "changelog"          "use_dependency"     "use_resource"      
[13] "multi-artefact"     "multifile-artefact" "changelog"          "html"               "interactive"        "use_resource_dir"  
[19] "connection"         "spaces"             "view"     
```

File: R/src.R [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+src_migrate_parameters <- function(cfg, dat) {
+  if (is.null(dat$parameters)) {
+    return(NULL)
+  }
+  ## TODO: there's some inconsistency here with pluralisation in
+  ## orderly3, might need some updating.
+  fmt <- "orderly3::orderly_parameters(%s)"
+  args <- sprintf("%s = %s", names(dat$parameters),
+                  vapply(dat$parameters, deparse, ""))
+  sprintf(fmt, paste(args, collapse = ", "))
```
Comment (r-ash) [inline R/src.R:149]: I think this is not formatting correctly, we're getting here for `default-param` where `dat$parameters` are

```Browse[2]> dat$parameters
$nmin
$nmin$default
[1] 0.5


$disease
$disease$default
[1] "HepB"
```

```
orderly3::orderly_parameters(nmin = list(default = 0.5), disease = list(default = \"HepB\"))
```

File: R/src.R [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+  ret <- character()
+  for (el in split(dat$depends, dat$depends$index)) {
+    name <- el$name[[1]]
+    query <- el$id[[1]]
+    there <- el$filename
+    here <- el$as
+    use <- sprintf("%s = %s", dquote_if_required(here), dquote(there))
+    if (length(there) == 1) {
+      str <- sprintf('orderly3::orderly_dependency("%s", "%s", c(%s))',
+                     name, query, use)
```
Comment (r-ash) [inline R/src.R:192]: If the query uses a parameter this is generating e.g.

```
orderly3::orderly_dependency("ago_data_areas", "latest(parameter:version == version)", c("depends/ago_areas.geojson" = "ago_areas.geojson"))
```

which doesn't work. Though I can't remember how are we supposed to pass in a parameter to the query and I can't see an example of that working in orderly3

File: tests/testthat/test-src.R (hunk 2 above)
Comment (richfitz) [reply tests/testthat/test-src.R:5]: yeah, this only runs the cases that are in the demo

File: R/src.R (hunk 4 above)
Comment (r-ash) [reply R/src.R:194]: https://github.com/mrc-ide/inference-data/blob/main/src/aaa_areas_pull/orderly.yml#L33

Comment (richfitz) [reply R/src.R:194]: Right, so this needs to become `latest(parameter:version == this:version)` I think...

**Review by r-ash — APPROVED** [review]

**Pull request merged**